agentlib/modules/simulation/csv_data_source.py

```diff
@@ -0,0 +1,160 @@
+from pathlib import Path
+from typing import List, Optional, Union
+
+import numpy as np
+import pandas as pd
+from pydantic import Field, field_validator, FilePath, model_validator
+
+from agentlib import Agent
+from agentlib.core import BaseModule, BaseModuleConfig, AgentVariable
+
+
+class CSVDataSourceConfig(BaseModuleConfig):
+    data: Union[pd.DataFrame, FilePath] = Field(
+        title="Data",
+        default=pd.DataFrame(),
+        description="Data that should be communicated during execution. Index should "
+        "be either numeric or Datetime, numeric values are interpreted as"
+        " seconds.",
+        validate_default=True,
+    )
+    outputs: List[AgentVariable] = Field(
+        title="Outputs",
+        default=[],
+        description="Optional list of columns of data frame that should be sent. If "
+        "ommited, all datapoint in frame are sent.",
+    )
+    t_sample: Union[float, int] = Field(
+        title="t_sample",
+        default=1,
+        ge=0,
+        description="Date source sends a value every <t_sample> seconds. Default is 1 s.",
+    )
+    data_offset: Optional[Union[pd.Timedelta, float]] = Field(
+        title="data_offset",
+        default=0,
+        description="Offset will be subtracted from index. I.e. if your environment "
+        "starts at 0, and you want your data-source to start at 1000 "
+        "seconds, you should set this to 1000.",
+    )
+    shared_variable_fields: List[str] = ["outputs"]
+
+    @field_validator("data")
+    @classmethod
+    def check_data(cls, data):
+        """Makes sure data is a data frame, and loads it if required."""
+        if isinstance(data, (str, Path)) and Path(data).is_file():
+            data = pd.read_csv(data, engine="python", index_col=0)
+        if not isinstance(data, pd.DataFrame):
+            raise ValueError(
+                f"Data {data} is not a valid DataFrame or the path is not found."
+            )
+        if data.empty:
+            raise ValueError("Provided data is empty.")
+        if len(data) < 2:
+            raise ValueError(
+                "The dataframe must contain at least two rows for interpolation."
+            )
+        return data
+
+    @model_validator(mode="after")
+    def are_outputs_in_data(self):
+        if self.outputs:
+            columns = set(self.data.columns)
+            output_names = set(o.name for o in self.outputs)
+
+            missing_columns = output_names - columns
+            if missing_columns:
+                raise ValueError(
+                    f"The following output columns are not present in the dataframe: "
+                    f"{', '.join(missing_columns)}"
+                )
+        return self
+
+
+class CSVDataSource(BaseModule):
+
+    config: CSVDataSourceConfig
+
+    def __init__(self, config: dict, agent: Agent):
+        super().__init__(config=config, agent=agent)
+
+        data = self.config.data
```

Comment:
Do you think it's beneficial? Since this code is not validation but working code, I thought its easier to understand and debug, if we keep this as regular python code outside of pydantic

Reply:
yes, could be. Only if a method fails it breaks the idea of "if the agent config is valid, it is able to run" we often follow

Reply:
I moved the index to numeric in the validator. The resampling is a bit much I think, as it changes the shape of the original dataframe (which is a config object) and doesnt raise any own Exceptions anyway.